waspc/src/Analyzer/Evaluator/Combinators.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | This module contains combinators for building evaluators for converting
+-- "TypedExpr" to arbitrary types.
+--
+-- A typical use:
+--
+-- @
+-- data Page = Page { title :: String, author :: Maybe String, content :: String }
+--
+-- page :: Evaluator Page
+-- page = dict $ Page <$> field "title" string <*> maybeField "author" string <*> field "content" string
+-- @
+--
+-- This creates a "Evaluator" @page@ that would turn the Wasp expression
+-- @{ title: "Home", content: "Hello world" }@ into
+-- @Page { title = "Home", author = Nothing, content = "Hello world" }@
+module Analyzer.Evaluator.Combinators
+  ( -- * Types
+    Evaluator,
+    DictEvaluator,
+
+    -- * Functions
+    build,
+
+    -- * "Evaluator" combinators
+    string,
+    integer,
+    double,
+    bool,
+    decl,
+    enum,
+    dict,
+    list,
+    extImport,
+    json,
+    psl,
+
+    -- * "DictEvaluator" combinators
+    field,
+    maybeField,
+  )
+where
+
+import Analyzer.Evaluator.Decl
+import Analyzer.Evaluator.EvaluationError
+import qualified Analyzer.Evaluator.Types as E
+import qualified Analyzer.Type as T
+import Analyzer.TypeChecker.AST (TypedExpr (..), exprType)
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Arrow (left)
+import Data.Functor.Compose (Compose (Compose, getCompose))
+import qualified Data.HashMap.Strict as H
+import Data.Typeable (cast)
+
+-- | The context in an evaluation.
+type EvalCtx a = (TD.TypeDefinitions, H.HashMap String Decl, a)
+
+-- | An evaluation from "a" to "b" with the evaluation context.
+type (|>) a b = Compose ((->) (EvalCtx a)) (Either EvaluationError) b
+
+-- | An evaluation from a typed expression to a value.
+newtype Evaluator a = Evaluator (TypedExpr |> a)
```

Comment:
> Also, the benefits of Reader are not really realized since the context is inspected in every combinator.

True, but in most combinators only "value" part of context is inspected, and "value" sounds like it shouldn't even be part of the context? So if "value" would be taken out of the context, then Reader would make sense?

I will not push this too much as I feel you have a much better grasp of this than me, so if you say Compose is a better solution here, I believe it. The only thing I was worried about is that more people know about monad transformers than Compose so Compose might be surprising. And we are having a lot of those `(_ _ smth)` in the code. But as I said, I don't really grasp the importance of combinators being applicative and so on, I don't feel I understand the problem good enough yet to be able to comment on that, so it is probably best if we continue with Compose.

I am still interested though, would it make any sense to try and remove that `a` from Context? Would that be an improvement or create a mess?

Comments explaining stuff -> sure, they can only help!

Reply:
Done with the PR.